keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,532 @@
+    - [Logger implementation performance](#logger-implementation-performance)
+    - [Log volume increase analysis](#log-volume-increase-analysis)
+  - [Future work](#future-work)
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Migration being abandoned halfway](#migration-being-abandoned-halfway)
+    - [Huge increase of log volume](#huge-increase-of-log-volume)
+- [Design Details](#design-details)
+  - [Proposed list of log messages to change](#proposed-list-of-log-messages-to-change)
+  - [Example of migrating klog call](#example-of-migrating-klog-call)
+- [Alternatives](#alternatives)
+    - [Just write guideline and update log messages](#just-write-guideline-and-update-log-messages)
+    - [Replace klog with some other structured logging library](#replace-klog-with-some-other-structured-logging-library)
+    - [Use glogr instead of proposed message structure](#use-glogr-instead-of-proposed-message-structure)
+- [Code organisation](#code-organisation)
+- [Production Readiness Review Process](#production-readiness-review-process)
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes to define standard structure for Kubernetes log messages, add methods to klog to enforce this structure, add ability to configure Kubernetes components to produce logs in JSON format and initiate migration to structured logging.
+
+## Motivation
+
+Current logging in the Kubernetes control plane doesn’t guarantee any uniform structure for log messages and references to Kubernetes objects in those logs. This makes parsing, processing, storing, querying and analyzing logs hard and forces administrators and developers to rely on ad-hoc solutions in most cases based on some regular expressions. Due to those problems any analytical solution based on those logs is hard to implement and maintain.
+
+### Goals
+
+* Make most common logs more queryable by standardizing log message and references to Kubernetes objects (Pods, Nodes etc.)
+* Enforce log structure by introduction of new klog methods that could be used to generate structured logs.
+* Propose reasonable scope of migration:
+  * Executable within SIG-instrumentation resources
+  * Minimize review burden on other SIGs
+  * Leave logging in acceptable state if effort fails
+* Simplify ingestion of logs into third party logging solutions by adding an option to output logs in the JSON format
+
+
+### Non-Goals
+
+* We are **not** replacing currently used logging library (klog) or the way in which it is used
+* We are **not** preserving full compatibility of logging output
+* We are **not** proposing structuring all logs in Kubernetes
+
+## Proposal
+
+This KEP proposes introducing a standardized log message structure. To enforce a new log message structure we are proposing adding new methods to the klog library: `InfoS`, `ErrorS` which will provide a structured interface for building log messages and creating new helper methods which will provide consistent identification of Kubernetes objects in logs.
```

Comment:
We wanted to avoid confusion with plural form. There was also concerns that difference between `Info` and `Infos` is hard to spot, but function signature is the same.